File: fake-service/features/step_definitions/mail_view.rb
A_MAIL = /[^\s@]+@[^\s@]+\.[^\s@]+/

Then(/^I see the mail has a cc and a bcc recipient$/) do
  within('.msg-header') do
    first('.cc').text.should =~ A_MAIL
    first('.bcc').text.should =~ A_MAIL
  end
end

Then(/^that email has the '(.*)' tag$/) do |tag|
  within('#mail-view') do |e|
    all('.tagsArea .tag').map(&:text).map(&:downcase).to_a.should include(tag)
  end
end

When(/I add the tag '(.*)' to that mail/) do |tag|
  page.execute_script("$('#new-tag-button').click();")
  page.execute_script("$('#new-tag-input').val('#{tag}');")
  find('#new-tag-input').native.send_keys [:return]
end

And(/^I reply to it$/) do
  click_button('Reply')
  click_button('Send')
end

Then(/^I choose to forward this mail$/) do
  click_button('Forward')
end

Then(/^I forward this mail$/) do
  click_button('Send')
end


Then(/^I remove all tags$/) do
  within('.tagsArea') do
    all('.tag').each do |tag|
      tag.click
    end
  end
end

Then(/^I choose to trash$/) do
  click_button('Trash message')
end

When(/^I try to delete the first mail$/) do
  step 'I open the first mail in the mail list'
  within('#mail-view') do
    page.driver.find_css('#view-more-actions')[0].click
    page.driver.execute_script("$('#delete-button-top').click();")
  end
  find('#user-alerts').text.should == 'Your message was moved to trash!'
end

Then(/^I see that the subject reads '(.*)'$/) do |expected_subject|
  find('#mail-view .subject').text.should == expected_subject
end

Then(/^I see that the body reads '(.*)'$/) do |expected_body|
  find('#mail-view .bodyArea').text.should == expected_body
end

Then(/^I see if the mail has html content/) do
  find('#mail-view .bodyArea').should have_css('h2[style*=\'color: #3f4944\']', :text => "cborim")
end

File: service/test/functional/features/steps/mail_view.py
#
# Copyright (c) 2014 ThoughtWorks, Inc.
#
# Pixelated is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Pixelated is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Pixelated. If not, see <http://www.gnu.org/licenses/>.

from behave import then, when
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.by import By

from common import (
    click_button,
    find_element_by_css_selector,
    find_elements_by_css_selector,
    reply_subject,
    wait_until_button_is_visible,
    wait_until_element_is_visible_by_locator)


@then('I see that the subject reads \'{subject}\'')
def impl(context, subject):
    e = find_element_by_css_selector(context, '#mail-view .mail-read-view__header-subject')
    assert e.text == subject


@then('I see that the body reads \'{expected_body}\'')
def impl(context, expected_body):
    find_element_by_css_selector(context, '#read-sandbox')
    context.browser.switch_to_frame('read-sandbox')
    e = find_element_by_css_selector(context, 'body')
    assert e.text == expected_body
    context.browser.switch_to_default_content()


@then('that email has the \'{tag}\' tag')
def impl(context, tag):
    wait_until_element_is_visible_by_locator(context, (By.CSS_SELECTOR, '#mail-view .tagsArea .tag'))
    elements = find_elements_by_css_selector(context, '#mail-view .tagsArea .tag')
    tags = [e.text for e in elements]
    assert tag in tags


@when('I add the tag \'{tag}\' to that mail')
def impl(context, tag):
    b = wait_until_element_is_visible_by_locator(context, (By.ID, 'new-tag-button'))
    b.click()

    e = wait_until_element_is_visible_by_locator(context, (By.ID, 'new-tag-input'))
    e.send_keys(tag)
    e.send_keys(Keys.ENTER)
    wait_until_element_is_visible_by_locator(context, (By.CSS_SELECTOR, 'li[data-tag=%s]' % tag))


@when('I reply to it')
def impl(context):
    click_button(context, 'Reply')
    context.reply_subject = reply_subject(context)
    click_button(context, 'Send')


@when('I try to delete the first mail')
def impl(context):
    context.execute_steps(u"When I open the first mail in the mail list")
    find_element_by_css_selector(context, '#mail-view #view-more-actions').click()
    context.browser.execute_script("$('#delete-button-top').click();")

    e = find_element_by_css_selector(context, '.message-panel__growl--success')
    assert 'Your message was moved to trash!' == e.text


@when('I choose to forward this mail')
def impl(context):
    wait_until_button_is_visible(context, 'Forward')
    click_button(context, 'Forward')


@when('I forward this mail')
def impl(context):
    wait_until_button_is_visible(context, 'Send')
    click_button(context, 'Send')


@when('I remove all tags')
def impl(context):
    e = find_element_by_css_selector(context, '.tagsArea')
    tags = e.find_elements_by_css_selector('.tag')
    assert len(tags) > 0
    for tag in tags:
        tag.click()


@when('I choose to trash')
def impl(context):
    context.browser.execute_script("$('button#view-more-actions').click()")
    click_button(context, 'Delete this message', 'li')


@then('I see the mail has a cc and a bcc recipient')
def impl(context):
    cc = find_element_by_css_selector(context, '.mail-read-view__header-recipients .cc')
    bcc = find_element_by_css_selector(context, '.mail-read-view__header-recipients .bcc')

    assert cc is not None
    assert bcc is not None
